### PR TITLE
26.2

### PR DESCRIPTION
**Pull request by margarita-ignatieva**

### CODE REVIEW

**Review by floriangeyer21 — APPROVED** [review]

File: src/main/java/com/internet/shop/controllers/order/AddOrderController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.internet.shop.library.Injector;
+import com.internet.shop.service.interfaces.OrderService;
+import com.internet.shop.service.interfaces.ShoppingCartService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AddOrderController extends HttpServlet {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/order/AddOrderController.java]:
```suggestion
public class CompleteOrderController extends HttpServlet {
```

File: src/main/java/com/internet/shop/controllers/order/AddOrderController.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    private final ShoppingCartService shoppingCartService =
+            (ShoppingCartService) injector.getInstance(ShoppingCartService.class);
+    private final OrderService orderService =
+            (OrderService) injector.getInstance(OrderService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/order/userOrders.jsp").forward(req, resp);
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/order/AddOrderController.java]: where do you use this?

File: src/main/webapp/WEB-INF/views/order/all.jsp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            <th>name</th>
+            <th>price</th>
+        </tr>
+        <c:forEach var="product" items="${order.products}">
+            <tr>
+                <td>${product.id}</td>
+                <td>${product.name}</td>
+                <td>${product.price}</td>
+            </tr>
+        </c:forEach>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/order/all.jsp]: let's have a button `details` instead that will do to OrderDetailsController and show a separate JSP with info about products

File: src/main/webapp/WEB-INF/views/order/userOrders.jsp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    </tr>
+    <c:forEach var="order" items="${orders}">
+        <tr>
+            <td>
+                <c:out value="${order.id}"/>
+            </td>
+            <td>
+                <a href="${pageContext.request.contextPath}
+                /orders/order?id=${order.id}">order details</a>
+            </td>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/order/userOrders.jsp]: like here
but it seems you haven't created the controller

File: src/main/webapp/web.xml [hunk 5]
```diff
@@ -35,43 +35,69 @@
...
+        <url-pattern>/user/orders</url-pattern>
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>getAllOrders</servlet-name>
+        <servlet-class>com.internet.shop.controllers.order.GetAllOrdersController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>getAllOrders</servlet-name>
+        <url-pattern>/orders/all</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/orders</url-pattern>
```

it seems like not all your new controllers have mapping in pom.xml

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
1. you need to create OrderDetails controller
2. let's add this functionality: 

> Создать две страницы с товарами:
> на одной можно только добавлять товар в корзину (уже реализована в HW 16.1),
> на другой (функционал для администратора сайта) отобразить список с возможностью только удалять товары а также кнопка для перехода на страницу создания нового товара.

3. You are merging your changes into the previous branch - this is wrong. You should have `master` branch in your project and merge all your changes into `master`. We expect you to create this branch and create a PR for merging all the changes into `master`.

File: src/main/java/com/internet/shop/controllers/cart/AddProductToCartController.java [hunk 6]
```diff
@@ -26,6 +26,6 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp)
         String productId = req.getParameter("productId");
         Product currentProduct = productService.get(Long.valueOf(productId));
         shoppingCartService.addProduct(shoppingCart, currentProduct);
-        resp.sendRedirect(req.getContextPath() + "/product/all");
+        resp.sendRedirect(req.getContextPath() + "shopping-cart/products/add");
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/cart/AddProductToCartController.java]: is this a kind of recursion?

File: src/main/java/com/internet/shop/controllers/cart/DeleteProductFromCartController.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            = (ShoppingCartService) injector.getInstance(ShoppingCartService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        ShoppingCart shoppingCart = shoppingCartService.getByUserId(USER_ID);
+        String productId = req.getParameter("productId");
+        shoppingCartService.deleteProduct(shoppingCart,
+                productService.get(Long.parseLong(productId)));
+        resp.sendRedirect(req.getContextPath() + "/shopping-cart/products/add");
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/cart/DeleteProductFromCartController.java]:
```suggestion
        resp.sendRedirect(req.getContextPath() + "/shopping-cart/products/");
```

File: src/main/java/com/internet/shop/controllers/order/CompleteOrderController.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            (ShoppingCartService) injector.getInstance(ShoppingCartService.class);
+    private OrderService orderService =
+            (OrderService) injector.getInstance(OrderService.class);
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        ShoppingCart shoppingCart = shoppingCartService.getByUserId(USER_ID);
+        Order order = orderService.completeOrder(shoppingCart);
+        order.setUserId(shoppingCart.getUserId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/order/CompleteOrderController.java]: isn't it done already inside completeOrder()?

File: src/main/webapp/WEB-INF/views/order/all.jsp (hunk 3 above)
Comment (KseniiaMakarova) [reply src/main/webapp/WEB-INF/views/order/all.jsp]: please fix this comment

File: src/main/webapp/WEB-INF/views/order/userOrders.jsp [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+             <th>price</th>
+         </tr>
+         <c:forEach var="product" items="${order.products}">
+             <tr>
+                 <td>${product.id}</td>
+                 <td>${product.name}</td>
+                 <td>${product.price}</td>
+             </tr>
+         </c:forEach>
+     </table>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/order/userOrders.jsp]: same - you should have `details` in front of each orderId
which would redirect you to OrderDetailsController and then to order/details.jsp

File: src/main/webapp/web.xml [hunk 10]
```diff
@@ -30,48 +31,74 @@
...
         <url-pattern>/shopping-cart/products</url-pattern>
     </servlet-mapping>
     <servlet>
         <servlet-name>addProduct</servlet-name>
-        <servlet-class>com.internet.shop.controllers.AddProductController</servlet-class>
+        <servlet-class>com.internet.shop.controllers.product.AddProductController</servlet-class>
     </servlet>
     <servlet-mapping>
         <servlet-name>addProduct</servlet-name>
         <url-pattern>/product/add</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/products/add</url-pattern>
```

File: src/main/webapp/web.xml [hunk 11]
```diff
@@ -30,48 +31,74 @@
...
         <url-pattern>/user/delete</url-pattern>
     </servlet-mapping>
     <servlet>
         <servlet-name>allProducts</servlet-name>
-        <servlet-class>com.internet.shop.controllers.GetAllProductsController</servlet-class>
+        <servlet-class>com.internet.shop.controllers.product.GetAllProductsController</servlet-class>
     </servlet>
     <servlet-mapping>
         <servlet-name>allProducts</servlet-name>
         <url-pattern>/product/all</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/products</url-pattern>
```

File: src/main/webapp/web.xml [hunk 12]
```diff
@@ -30,48 +31,74 @@
...
     </servlet-mapping>
 
     <servlet>
         <servlet-name>deleteUser</servlet-name>
-        <servlet-class>com.internet.shop.controllers.DeleteUserController</servlet-class>
+        <servlet-class>com.internet.shop.controllers.user.DeleteUserController</servlet-class>
     </servlet>
     <servlet-mapping>
         <servlet-name>deleteUser</servlet-name>
         <url-pattern>/user/delete</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/users/delete</url-pattern>
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — COMMENTED** [review]
1. you need to create OrderDetails controller
2. let's add this functionality:

> Создать две страницы с товарами:
> на одной можно только добавлять товар в корзину (уже реализована в HW 16.1),
> на другой (функционал для администратора сайта) отобразить список с возможностью только удалять товары а также кнопка для перехода на страницу создания нового товара.

File: src/main/java/com/internet/shop/controllers/order/CompleteOrderController.java (hunk 8 above)
Comment (margarita-ignatieva) [reply src/main/java/com/internet/shop/controllers/order/CompleteOrderController.java]: yes, no need to create order

File: src/main/webapp/WEB-INF/views/order/all.jsp [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        <tr>
+            <th>name</th>
+        </tr>
+        <c:forEach var="product" items="${order.products}">
+            <tr>
+                <td>${product.name}</td>
+            </tr>
+        </c:forEach>
+        <a href="${pageContext.request.contextPath}/orders/delete?orderId=${order.id}">delete</a>
+    </table>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/order/all.jsp]: make a table with 4 columns:

Order Id
User Id
Order details
Delete order

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
you should have 2 controllers for showing products:

1. ManageProductsController - this is a page for admin.
Display all products with their names and prices in a table. Next to each product - `Delete` option that will delete the product from the storage.
Below the products table - `Add new product` button that redirects to adding a new product to the storage.

2. GetAllProductsController - this is a page for user.
Display all products with their names and prices in a table. Next to each product - `Buy` option that will add the product to the shopping cart.
Below the products table - `Go to cart` button that redirects to showing the shopping cart.

File: src/main/webapp/WEB-INF/views/order/all.jsp (hunk 13 above)
Comment (margarita-ignatieva) [reply src/main/webapp/WEB-INF/views/order/all.jsp]: Added jsp for admin and user, not sure about interactions for them inside project

File: src/main/webapp/WEB-INF/views/order/all.jsp [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        </tr>
+        <c:forEach var="order" items="${orders}">
+            <tr>
+                <td><c:out value="${order.orderId}"/></td>
+            </tr>
+            <tr>
+                <td><c:out value="${order.userId}"/></td>
+            </tr>
+            <tr>
+                <a href="${pageContext.request.contextPath}/order/details?orderID=${order.id}">details</a>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/order/all.jsp]:
```suggestion
                <a href="${pageContext.request.contextPath}/order/details?orderId=${order.id}">details</a>
```

File: src/main/java/com/internet/shop/controllers/order/GetOrderDetailsController.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+public class GetOrderDetailsController extends HttpServlet {
+    private static final Injector injector = Injector.getInstance("com.internet.shop");
+    private OrderService orderService = (OrderService) injector
+            .getInstance(OrderService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        Long id = Long.valueOf(req.getParameter("productId"));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/order/GetOrderDetailsController.java]:
```suggestion
        Long id = Long.valueOf(req.getParameter("orderId"));
```

File: src/main/webapp/WEB-INF/views/order/userOrders.jsp [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        <th>Order id</th>
+    </tr>
+
+    <c:forEach var="order" items="${orders}">
+        <tr>
+            <td>
+                <c:out value="${order.orderId}"/>
+            </td>
+            <td>
+                <a href="${pageContext.request.contextPath}/order/details?id=${order.orderId}">Order details</a>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/order/userOrders.jsp]:
```suggestion
                <a href="${pageContext.request.contextPath}/order/details?orderId=${order.orderId}">Order details</a>
```

File: src/main/webapp/WEB-INF/views/product/adminAll.jsp [hunk 17]
```diff
@@ -27,7 +22,10 @@
                 <c:out value="${product.price}"/>
             </td>
             <td>
-                <a href = "${pageContext.request.contextPath}/products/add?id=${product.id}"> Buy </a>
+                <a href="${pageContext.request.contextPath}/products/delete?id=${product.id}">Delete</a>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/product/adminAll.jsp:25]: I can't see this controller in your PR

File: src/main/webapp/WEB-INF/views/shoppingCart/allProducts.jsp [hunk 18]
```diff
@@ -24,7 +24,7 @@
                 <c:out value="${product.price}"/>
             </td>
             <td>
-                <a href = "${pageContext.request.contextPath}/cart/delete?id=${product.id}"> Delete </a>
+                <a href = "${pageContext.request.contextPath}/shopping-cart/products/add?id=${product.id}"> Delete </a>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/shoppingCart/allProducts.jsp]:
```suggestion
                <a href = "${pageContext.request.contextPath}/shopping-cart/products/delete?id=${product.id}"> Delete </a>
```

File: src/main/webapp/web.xml [hunk 19]
```diff
@@ -30,48 +31,105 @@
...
-        <servlet-class>com.internet.shop.controllers.AddProductToCartController</servlet-class>
+        <servlet-name>deleteProduct</servlet-name>
+        <servlet-class>com.internet.shop.controllers.cart.DeleteProductFromCartController</servlet-class>
     </servlet>
+
     <servlet-mapping>
-        <servlet-name>buyProduct</servlet-name>
-        <url-pattern>/shopping-cart/products/add</url-pattern>
+        <servlet-name>deleteProduct</servlet-name>
+        <url-pattern>/products/delete</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:71]:
```suggestion
        <url-pattern>/shopping-cart/products/delete</url-pattern>
```
add one more controller with mapping `/products/delete` for deletion of product from storage (admin does this)

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**